[fix] Improve error message if `this` attribute of `<svelte:component>` is not valid

baseballyama opened the pull request:
fix: https://github.com/sveltejs/svelte/issues/7550

I added better error message in dev mode.
And I improved error message of SSR little bit. 

Improved error message is like below.

<img width="592" alt="image" src="https://user-images.githubusercontent.com/19153718/169636379-0922aa15-ff22-41bc-becf-35c69f2085c4.png">

### Before submitting the PR, please make sure you do the following
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] Prefix your PR title with `[feat]`, `[fix]`, `[chore]`, or `[docs]`.
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
-  [x] Run the tests with `npm test` and lint the project with `npm run lint`

Conduitry commented:
I don't think we can be this strict about the check. People might be using pre-compiled components - and even though that causes other things to watch out for, I don't think we should prevent it outright. Or they could even be using a Proxy wrapper around a component for some specialized behavior. I'm not sure what check we can perform here other than just to see whether it's a constructor.

UltraCakeBakery commented:
If only the `switch_value` part of the error in my original issue was `this={...} of <svelte:component>` instead, i would not have made the issue. Maybe there is no need to do an additional check?

_I have no idea how svelte works under the hood so feel free to ignore this comment if it was absolutely useless to y'all 😅_

baseballyama commented:
I updated the validation logic.
Instead of class validation (`if (!(component.prototype instanceof SvelteComponent))`),
I added contractor check and expected functions are there. (`if (!instance.$$ || !instance.$set || !instance.$on || !instance.$destroy)`)

Still, I'm not sure such a strict check is necessary at runtime. (As a Svelte user, I will try to avoid such errors by TypeScript type checking...)
But a better error message will increase developer experience, so once I updated it.

I think now we have 3 ways.
1. Don't do even contractor check
2. Just do contractor check
3. check both contractor and expected functions (`$set`, `$on`, `$destroy`)

Maybe I could be missing a reason why we shouldn't do 3, but what do you think @Conduitry ?

baseballyama commented on src/runtime/internal/dev.ts:
```diff
@@ -121,6 +121,24 @@ export function validate_void_dynamic_element(tag: undefined | string) {
+		return instance;
+	} catch (err) {
+		const { message } = err;
+		if (typeof message === 'string' && message.indexOf('is not a constructor') !== -1) {
```
This constructor check is messy but AFAIK there is no better way.

UltraCakeBakery commented:
> I updated the validation logic. Instead of class validation (`if (!(component.prototype instanceof SvelteComponent))`), I added contractor check and expected functions are there. (`if (!instance.$$ || !instance.$set || !instance.$on || !instance.$destroy)`)
> 
> Still, I'm not sure such a strict check is necessary at runtime. (As a Svelte user, I will try to avoid such errors by TypeScript type checking...) But a better error message will increase developer experience, so once I updated it.
> 
> I think now we have 3 ways.
> 
> 1. Don't do even contractor check
> 2. Just do contractor check
> 3. check both contractor and expected functions (`$set`, `$on`, `$destroy`)
> 
> Maybe I could be missing a reason why we shouldn't do 3, but what do you think @Conduitry ?

Are these error message checks also included in the final build? Or only during dev?

baseballyama commented:
> Are these error message checks also included in the final build? Or only during dev?

only during dev!
One of the important points of Svelte is performance.
Therefore I don't want to add such check logic in prod.

UltraCakeBakery commented:
> > Are these error message checks also included in the final build? Or only during dev?
> 
> only during dev! One of the important points of Svelte is performance. Therefore I don't want to add such check logic in prod.

Cool! I figured but I thought I would ask to make sure maintainers do not ask this right before they plan to finally merge this haha
